genie-web/src/main/java/com/netflix/genie/web/agent/apis/rpc/v4/endpoints/GRpcJobKillServiceImpl.java

```diff
@@ -204,14 +205,36 @@ public void killJob(
     @Scheduled(fixedDelay = 30_000L, initialDelay = 30_000L)
     public void cleanupOrphanedObservers() {
         for (final String jobId : this.parkedJobKillResponseObservers.keySet()) {
-            if (!this.agentRoutingService.isAgentConnectionLocal(jobId)) {
-                final StreamObserver<JobKillRegistrationResponse> observer = this.parkedJobKillResponseObservers.remove(
-                    jobId
-                );
-                if (observer != null) {
-                    observer.onCompleted();
+            try {
+                if (!this.agentRoutingService.isAgentConnectionLocal(jobId)) {
+                    final StreamObserver<JobKillRegistrationResponse> observer = this.parkedJobKillResponseObservers
+                        .remove(jobId);
+
+                    if (observer != null && !isStreamObserverCancelled(observer)) {
```

Comment:
You could move the null check and all this logic into your function and just call it `cancelObserverIfNecessary`

Reply:
Sounds good!